### ISSUE TITLE
CQT returns unexpected shape (number of columns)

### ISSUE THREAD

**Issue by bluenote10** [librosa/librosa#1002]
#### Description

I'm getting an unexpected output shape from CQT in case the audio length is an exact multiple of the hop length. For example, if the audio length is 1024 and the hop length is 512, I'm getting 3 CQT columns instead of 2.

I'm not sure if this is intended behavior, but at least it was quite unexpected and leading to a bug in my code.

#### Steps/Code to Reproduce

```python
import librosa
import numpy as np

rand_sample_1023 = np.random.uniform(low=-1.0, high=+1.0, size=1023)
rand_sample_1024 = np.random.uniform(low=-1.0, high=+1.0, size=1024)

C = librosa.core.cqt(rand_sample_1023, hop_length=512)
assert C.shape[1] == 2  # that's fine, 1023 allows for 2 columns

C = librosa.core.cqt(rand_sample_1024, hop_length=512)
assert C.shape[1] == 2  # violated, 3 columns now, even though 2 columns would still fit
```

#### Expected Results

Assertion passes.

#### Actual Results

Assertion fails.

#### Versions

```
Linux-4.15.0-65-generic-x86_64-with-Ubuntu-18.04-bionic
('Python', '2.7.15rc1 (default, Apr 15 2018, 21:51:34) \n[GCC 7.3.0]')
('NumPy', '1.15.2')
('SciPy', '1.1.0')
('librosa', '0.7.0')
```

**Comment by lostanlen** [comment]
FYI, the same arises in `stft` (which is called by `cqt`):

```
In [1]: librosa.stft(rand_sample_1023).shape
Out[1]: (1025, 2)

In [2]: librosa.stft(rand_sample_1024).shape
Out[2]: (1025, 3)
```

This is due to half-frame padding for centering the analysis window:
https://github.com/librosa/librosa/blob/b1e41e83a9d177d7ca83bf7e0160c5e3c4177e3b/librosa/core/spectrum.py#L217

The rationale is that, if the input waveform is 1023 samples long, then you can measure its cqt at two moments separated by `hop_length` (namely, 0 and 512), but if the waveform is 1024 samples long, then you can measure its cqt at three moments separated by `hop_length` (0, 512, and 1024).

TL;DR i think that this is intended behavior, and that no change is needed.

I hope this helps!

**Comment by lostanlen** [comment]
Related: #600

**Comment by bluenote10** [comment]
> if the waveform is 1024 samples long, then you can measure its cqt at three moments separated by hop_length (0, 512, and 1024).

That's exactly what goes against common sense: With zero-indexing a 1024 long vector doesn't have an index 1024, that would require a vector of length 1025, where 3 columns would naturally make sense.

**Comment by bmcfee** [comment]
> If that is really the desired behavior: How exactly do these 1024 elements contribute to the three computations of 512 elements each?

As @lostanlen says, this is due to padding inside the STFT.  The first frame covers samples `[-n_fft//2, n_fft//2)` (centered at 0), the second frame covers samples `[-n_fft//2 + hop, n_fft//2 + hop)`, and the third covers samples `[-n_fft//2 + 2*hop, n_fft//2 + 2 * hop)`.  Using your numbers (and, let's say an FFT size of 1024), this results in frames:

- `[-512, 512)`
- `[0, 1024)`
- `[512, 1536)`

After that point, the next frame would start at index `1024` which has zero overlap with the original signal, so we stop there.

The situation is a bit trickier with CQT because the window length is variable, and the signal duration is also subject to change during down-sampling.  But to answer your original question, yes this is all intended behavior.
